### PR TITLE
implement integration tests for every scenario

### PR DESCRIPTION
**Pull request by mbaechler**

### CODE REVIEW

File: src/test/scala-2.12/com/linagora/gatling/imap/CyrusServer.scala [hunk 1]
```diff
@@ -18,22 +18,36 @@
...
-    def createUser(login: String, password: String): RunningCyrusServer = {
+
+    def addUser(login: String, password: String): Unit = {
       container.execInContainer("bash", "-c", s"echo $password | saslpasswd2 -u test -c $login -p")
-      this
+      implicit val executionContext: ExecutionContextExecutor = ExecutionContext.global
+      withConnectedSession (mappedImapPort) { implicit session =>
+        Await.result(for {
+          _ <- Imap.login("cyrus", "cyrus")
+          _ <- Imap.rawCommand("CREATE user.user1")
```
Comment (rouazana) [inline src/test/scala-2.12/com/linagora/gatling/imap/CyrusServer.scala]: you mean user.$login?

File: src/test/scala-2.12/com/linagora/gatling/imap/protocol/ImapSessionsSpec.scala [hunk 2]
```diff
@@ -16,25 +17,26 @@ class ImapSessionsSpec extends WordSpec with Matchers with ImapTestUtils with Be
...
       val protocol = ImapProtocol("localhost", port = server.mappedImapPort(), config = config)
 
       val sessions = system.actorOf(ImapSessions.props(protocol))
       val probe = TestProbe()
-      probe.send(sessions, Command.Connect("1"))
+      val userId = "1"
+      probe.send(sessions, Command.Connect(userId))
       probe.expectMsg(10.second, Response.Connected(ImapResponses.empty))
-      probe.send(sessions, Command.Login("1", "user1", "password"))
+      probe.send(sessions, Command.Login(userId, bart))
```
Comment (rouazana) [inline src/test/scala-2.12/com/linagora/gatling/imap/protocol/ImapSessionsSpec.scala]: what's the purpose of this userId?

File: src/it/resources/logback.xml [hunk 3]
```diff
@@ -2,11 +2,11 @@
 
     <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
         <encoder>
-            <pattern>%d{HH:mm:ss.SSS} [%thread] %-5level %logger{36} - %msg%n</pattern>
+            <pattern>%d{HH:mm:ss.SSS} %-5level %logger{36} - %msg%n</pattern>
```
Comment (rouazana) [inline src/it/resources/logback.xml]: why?

File: src/it/resources/logback.xml [hunk 4]
```diff
@@ -2,11 +2,11 @@
...
         </encoder>
     </appender>
 
-    <root level="ERROR">
+    <root level="TRACE">
```
Comment (rouazana) [inline src/it/resources/logback.xml]: does it finally work?

File: src/it/scala-2.12/logback-test.xml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<configuration>
+
+    <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
+        <encoder>
+            <pattern>--------------- %d{HH:mm:ss.SSS} [%thread] %-5level %logger{36} - %msg%n</pattern>
```
Comment (rouazana) [inline src/it/scala-2.12/logback-test.xml]: debug? :)

File: src/main/scala-2.12/com/linagora/gatling/imap/Imap.scala [hunk 6]
```diff
@@ -26,7 +26,9 @@ object Imap {
     command(new IMAPCommandListener {
       override def onResponse(session: IMAPSession, tag: String, responses: util.List[IMAPResponse]): Unit = {
         import collection.JavaConverters._
-        taggedResponse.success(TaggedResponse(tag, responses.asScala.toList))
+        val r = responses.asScala.toList
+        println(s"cmd response : $r")
+        taggedResponse.success(TaggedResponse(tag, r))
```
Comment (rouazana) [inline src/main/scala-2.12/com/linagora/gatling/imap/Imap.scala]: debug?

File: src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapAuthenticationScenario.scala [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+import io.gatling.core.feeder.FeederBuilder
+import io.gatling.core.structure.ScenarioBuilder
+
+import scala.collection.immutable.Seq
+
+object ImapAuthenticationScenario {
+
+  def apply(feeder: FeederBuilder): ScenarioBuilder = scenario("ImapAuthentication")
+    .feed(feeder)
+    .exec(imap("Connect").connect())//.exitHereIfFailed
```
Comment (rouazana) [inline src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapAuthenticationScenario.scala]: what's this comment?

File: src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapStoreScenario.scala [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    Validation.sequence(value.flags.map(flag => flag.el[String]).map(expr => expr(session)))
+      .map(mutable => collection.immutable.Seq(mutable:_*))
+      .map(flags => value.setFlags(flags))
+
+  private val rangeFlagsUpdates = imap("storeAll").store(MessageRanges(From(1L)), StoreFlags.add(Silent.Enable(), "all${loopId}")).check(ok)
+  private val singleFlagsUpdate = imap("storeOne").store(session => MessageRanges(One(1L + Random.nextInt(numberOfMailInInbox))), StoreFlags.add(Silent.Enable(), "one${loopId}")).check(ok)
+
+  def apply(feeder: FeederBuilder): ScenarioBuilder =
+    scenario("imap store scenario")
+    .feed(feeder)
```
Comment (rouazana) [inline src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapStoreScenario.scala]: indent

File: src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapUIDFetchScenario.scala [hunk 9]
```diff
@@ -26,17 +26,18 @@ class ImapUIDFetchScenario extends Simulation {
...
+  private val populateInbox = repeat(numberOfMailInInbox)(pause(appendGracePeriod).exec(populateMailbox))
 
-  val uidFetch = exec(imap("uidFetch").uidFetch(MessageRanges(Range(1, numberOfMailInInbox)), AttributeList("UID")).check(ok))
+  private val uidFetch = exec(imap("uidFetch").uidFetch(MessageRanges(Range(1, numberOfMailInInbox)), AttributeList("UID")).check(ok))
   
-  val scn = scenario("Imap").feed(feeder)
+  def apply(feeder: FeederBuilder): ScenarioBuilder =
+    scenario("Imap")
+      .feed(feeder)
     .exec(imap("Connect").connect()).exitHereIfFailed
```
Comment (rouazana) [inline src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapUIDFetchScenario.scala]: indent

File: src/main/scala-2.12/logback.xml [hunk 10]
```diff
@@ -0,0 +1,12 @@
+<configuration>
```
Comment (rouazana) [inline src/main/scala-2.12/logback.xml]: 3 logback files! :)

Comment (remk) [reply src/main/scala-2.12/logback.xml]: This one could be deleted, they shoud be in the resources directory

File: src/test/scala-2.12/com/linagora/gatling/imap/CyrusServer.scala (hunk 1 above)
Comment (mbaechler) [reply src/test/scala-2.12/com/linagora/gatling/imap/CyrusServer.scala]: yes, fixed in a later commit

File: src/test/scala-2.12/com/linagora/gatling/imap/protocol/ImapSessionsSpec.scala (hunk 2 above)
Comment (mbaechler) [reply src/test/scala-2.12/com/linagora/gatling/imap/protocol/ImapSessionsSpec.scala]: if I remember correctly, it creates an actor with this name to handle messages linked to a "session"

Comment (rouazana) [reply src/test/scala-2.12/com/linagora/gatling/imap/protocol/ImapSessionsSpec.scala]: would you mind strong typing it?

File: src/it/resources/logback.xml (hunk 3 above)
Comment (mbaechler) [reply src/it/resources/logback.xml]: I think that in an actor model, it doesn't bring much value to know the thread we are running and it makes the logs smaller

File: src/it/resources/logback.xml (hunk 4 above)
Comment (mbaechler) [reply src/it/resources/logback.xml]: yes

File: src/it/scala-2.12/logback-test.xml (hunk 5 above)
Comment (mbaechler) [reply src/it/scala-2.12/logback-test.xml]: yes, removed

File: src/main/scala-2.12/com/linagora/gatling/imap/Imap.scala (hunk 6 above)
Comment (mbaechler) [reply src/main/scala-2.12/com/linagora/gatling/imap/Imap.scala]: yes

File: src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapAuthenticationScenario.scala (hunk 7 above)
Comment (mbaechler) [reply src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapAuthenticationScenario.scala]: some debug left

File: src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapStoreScenario.scala (hunk 8 above)
Comment (mbaechler) [reply src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapStoreScenario.scala]: ok

File: src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapUIDFetchScenario.scala (hunk 9 above)
Comment (mbaechler) [reply src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapUIDFetchScenario.scala]: ok

File: src/test/scala-2.12/com/linagora/gatling/imap/protocol/ImapSessionsSpec.scala (hunk 2 above)
Comment (mbaechler) [reply src/test/scala-2.12/com/linagora/gatling/imap/protocol/ImapSessionsSpec.scala]: I can

File: src/main/scala-2.12/com/linagora/gatling/imap/Imap.scala (hunk 6 above)
Comment (rouazana) [reply src/main/scala-2.12/com/linagora/gatling/imap/Imap.scala]: you don't need this variable extraction anymore then

File: src/test/scala-2.12/com/linagora/gatling/imap/protocol/ImapSessionsSpec.scala (hunk 2 above)
Comment (mbaechler) [reply src/test/scala-2.12/com/linagora/gatling/imap/protocol/ImapSessionsSpec.scala]: Done. Actually, it comes from gatling and it's supposed to be a Long.

File: src/main/scala-2.12/com/linagora/gatling/imap/Imap.scala (hunk 6 above)
Comment (mbaechler) [reply src/main/scala-2.12/com/linagora/gatling/imap/Imap.scala]: ok

**Pull request merged**